packages/vm/src/memory.rs

```diff
@@ -122,29 +122,43 @@ pub fn write_region(ctx: &Ctx, ptr: u32, data: &[u8]) -> VmResult<()> {
                 cells[i].set(data[i])
             }
             region.length = data.len() as u32;
-            set_region(ctx, ptr, region);
+            set_region(ctx, ptr, region)?;
             Ok(())
         },
-        None => panic!(
+        None => Err(CommunicationError::deref_err(region.offset, format!(
             "Error dereferencing region {:?} in wasm memory of size {}. This typically happens when the given pointer does not point to a Region struct.",
             region,
             memory.size().bytes().0
-        ),
+        )).into()),
     }
 }
 
 /// Reads in a Region at ptr in wasm memory and returns a copy of it
-fn get_region(ctx: &Ctx, ptr: u32) -> Region {
+fn get_region(ctx: &Ctx, ptr: u32) -> CommunicationResult<Region> {
     let memory = ctx.memory(0);
     let wptr = WasmPtr::<Region>::new(ptr);
-    let cell = wptr.deref(memory).unwrap();
-    cell.get()
+    match wptr.deref(memory) {
```

Comment:
Would this not work well?
```rust
if let Some(cell) = wptr.defer(memory) {
  cell.get()
} else {
  Err(CommunicationError::deref_err(ptr, "Could not dereference this pointer to a Region")
}
```

Reply:
Yes, but for me the ideal flow would be to not treat both cases equally but return on error and continue in the main body otherwise

Reply:
like

```rust
if let None = wptr.defer(memory) {
  return Err(CommunicationError::deref_err(ptr, "Could not dereference this pointer to a Region")
}

// here we know the value is Some(cell) but we don't have it available unwrapped
```

Reply:
```rust
let cell = wptr.defer(memory)
  .ok_or_else(|| 
    Err(CommunicationError::deref_err(ptr, "Could not dereference this pointer to a Region"))
  )?;
Ok(cell.get())
```

Reply:
You could even try using [`fehler`](https://boats.gitlab.io/blog/post/failure-to-fehler/) if you really don't like the `Ok`-wrapping. I even like the idea, but i don't know if it's worth refactoring with it right now, as it'll take a while to apply to everywhere in the codebase (you don't have to, but for consistency)

Reply:
oh yeah, i never tried it but fehler would probably totally mess up IntelliJ's type inference :sweat_smile:

Reply:
This creates an intermediate Result we don't need just because we cannot use the `return` operator due to the lack of the guard statement.

I don't think there a great solution until we get better language support for those cases:
- https://github.com/rust-lang/rfcs/pull/1303
- https://github.com/rust-lang/rfcs/issues/2616

Reply:
It's likely that both get optimized to the same code, but if you want to be extra sure of that you can play around with https://godbolt.org.
I'd do it myself if i could read.


Assmebly, that it.

Reply:
Yeah, I'm not worried about the compiled code. But the complexity does not got any better even if you can safe 2 lines of code.

Reply:
All the approaches in this thread are valid in my opinion, including your original implementation, so i don't mind which one you choose to use here. Feel free to resolve this if you prefer the original implementation.

Reply:
Exactly, all of them are okay, none of them is perfect